File: internal/tui/repl.go
// Package tui is a minimal REPL front-end for the orchestrator agent.
package tui

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"google.golang.org/adk/agent"
	"google.golang.org/adk/runner"
	"google.golang.org/genai"
)

const (
	cyan  = "\033[36m"
	gray  = "\033[90m"
	reset = "\033[0m"
)

// compactArgs renders function-call arguments as compact JSON for a one-line
// trace, falling back to Go's default formatting if marshaling fails.
func compactArgs(args map[string]any) string {
	if len(args) == 0 {
		return ""
	}
	if b, err := json.Marshal(args); err == nil {
		return string(b)
	}
	return fmt.Sprintf("%v", args)
}

// renderWidget prints a structured widget the worker returned in an A2A
// DataPart. Invoked from the widget handler Run registers, so widgets reach the
// terminal directly instead of being flattened into the LLM's text answer. It
// dispatches on the "_kind" the client injected from the part's metadata.
func renderWidget(w map[string]any) {
	title, _ := w["title"].(string)
	fmt.Printf("\n%s┌─ %s%s\n", cyan, title, reset)
	switch w["_kind"] {
	case "widget/order":
		if o, ok := w["order"].(map[string]any); ok {
			field := func(label, key string) {
				if v, ok := o[key]; ok && v != nil && v != "" {
					fmt.Printf("%s│%s %-9s %v\n", cyan, reset, label, v)
				}
			}
			field("Товар:", "item")
			field("Статус:", "status_label")
			if amt, ok := o["amount"]; ok {
				fmt.Printf("%s│%s %-9s %v %v\n", cyan, reset, "Сумма:", amt, o["currency"])
			}
			field("Клиент:", "customer")
			field("Дата:", "created")
		}
	case "widget/order_list":
		if rows, ok := w["orders"].([]any); ok {
			for _, r := range rows {
				o, ok := r.(map[string]any)
				if !ok {
					continue
				}
				fmt.Printf("%s│%s #%v  %v — %v (%v %v, %v)\n", cyan, reset,
					o["id"], o["item"], o["status_label"], o["amount"], o["currency"], o["created"])
			}
		}
	default: // widget/confirmation and any future dialog-style widget
		if message, _ := w["message"].(string); message != "" {
			fmt.Printf("%s│%s %s\n", cyan, reset, message)
		}
		if labels := actionLabels(w["actions"]); labels != "" {
			fmt.Printf("%s│%s %s\n", cyan, reset, labels)
		}
	}
	fmt.Printf("%s└─%s\n", cyan, reset)
}

// actionLabels renders a widget's action buttons as "[Label]  [Label]".
func actionLabels(v any) string {
	actions, ok := v.([]any)
	if !ok {
		return ""
	}
	var labels []string
	for _, a := range actions {
		if m, ok := a.(map[string]any); ok {
			if l, ok := m["label"].(string); ok {
				labels = append(labels, "["+l+"]")
			}
		}
	}
	return strings.Join(labels, "  ")
}

// WidgetSource is anything that can push structured widgets to a handler —
// implemented by the A2A orders client. The TUI registers a handler so widgets
// render directly here instead of passing through the LLM.
type WidgetSource interface {
	SetWidgetHandler(func(sessionID string, w map[string]any))
}

// Run starts a REPL that reads user input, runs it through the orchestrator
// runner, and prints assistant responses. Widgets from ws render inline; when a
// widget is shown during a turn, the assistant's text echo is suppressed so the
// same prompt/result isn't displayed twice (widget + text). Returns nil on
// "exit"/"quit" or EOF.
func Run(ctx context.Context, r *runner.Runner, ws WidgetSource) error {
	fmt.Printf("%sАссистент службы заказов.%s Введите запрос или 'exit' для выхода.\n", cyan, reset)
	in := bufio.NewScanner(os.Stdin)
	in.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	const userID, sessID = "tui-user", "tui-session"

	// widgetShown flips when a widget renders during the current turn. Safe as a
	// plain bool: the handler fires synchronously on this goroutine (the tool
	// call runs inside r.Run's iteration), never concurrently.
	widgetShown := false
	if ws != nil {
		ws.SetWidgetHandler(func(_ string, w map[string]any) {
			renderWidget(w)
			widgetShown = true
		})
	}

	for {
		fmt.Printf("%sвы>%s ", cyan, reset)
		if !in.Scan() {
			return in.Err()
		}
		line := in.Text()
		if line == "exit" || line == "quit" || line == "выход" {
			return nil
		}
		if line == "" {
			continue
		}
		widgetShown = false
		msg := genai.NewContentFromText(line, genai.RoleUser)
		fmt.Printf("%s  · агент → LLM: запрос%s\n", gray, reset)
		for event, err := range r.Run(ctx, userID, sessID, msg, agent.RunConfig{}) {
			if err != nil {
				fmt.Printf("%s[ошибка] %v%s\n", gray, err, reset)
				break
			}
			if event == nil || event.Content == nil {
				continue
			}
			var combined string
			for _, p := range event.Content.Parts {
				switch {
				case p.FunctionCall != nil:
					// LLM решило, какой инструмент вызвать.
					fmt.Printf("%s  · LLM → агент: вызвать %s(%s)%s\n",
						gray, p.FunctionCall.Name, compactArgs(p.FunctionCall.Args), reset)
				case p.FunctionResponse != nil:
					// Результат инструмента уходит обратно в LLM.
					fmt.Printf("%s  · инструмент → LLM: результат %s, снова спрашиваю LLM%s\n",
						gray, p.FunctionResponse.Name, reset)
				case p.Text != "":
					combined += p.Text
				}
			}
			// Skip the text echo when a widget already conveyed this turn's result.
			if combined != "" && !widgetShown {
				fmt.Printf("%sассистент>%s %s\n", cyan, reset, combined)
			}
		}
	}
}

File: java/orchestrator/src/main/java/io/github/kmpavloff/a2ademo/orchestrator/tui/Repl.java
package io.github.kmpavloff.a2ademo.orchestrator.tui;

import io.github.kmpavloff.a2ademo.orchestrator.a2a.OrdersClient;
import io.github.kmpavloff.a2ademo.orchestrator.agent.OrchestratorAgent;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/**
 * Minimal line-oriented REPL (port of internal/tui/repl.go): widgets from the
 * worker render inline as cards; when a widget was shown this turn, a verbose
 * text echo is suppressed so the same data is not displayed twice.
 */
public class Repl {
    private static final String CYAN = "\u001B[36m";
    private static final String GRAY = "\u001B[90m";
    private static final String RESET = "\u001B[0m";

    private final OrchestratorAgent agent;
    private final OrdersClient orders;

    // Flips when a widget renders during the current turn. Safe as a plain
    // field: the handler fires synchronously on the REPL thread.
    private boolean widgetShown;

    public Repl(OrchestratorAgent agent, OrdersClient orders) {
        this.agent = agent;
        this.orders = orders;
    }

    public void run() throws IOException {
        System.out.printf("%sАссистент службы заказов.%s Введите запрос или 'exit' для выхода.%n", CYAN, RESET);
        orders.setWidgetHandler((sessionId, w) -> {
            renderWidget(w);
            widgetShown = true;
        });
        // Files (e.g. the refund receipt) are saved next to the REPL and the
        // path is printed, since a terminal cannot "download" anything.
        orders.setFileHandler((sessionId, filename, mediaType, data) -> {
            try {
                Path name = Path.of(Path.of(filename).getFileName().toString());
                Files.write(name, data);
                System.out.printf("%s💾 Квитанция сохранена: ./%s%s%n", CYAN, name, RESET);
            } catch (IOException | InvalidPathException e) {
                System.out.printf("%s[файл %s не сохранён: %s]%s%n", GRAY, filename, e.getMessage(), RESET);
            }
        });

        BufferedReader in = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
        final String sessionId = "tui-session";
        while (true) {
            System.out.printf("%sвы>%s ", CYAN, RESET);
            String line = in.readLine();
            if (line == null || line.equals("exit") || line.equals("quit") || line.equals("выход")) {
                return;
            }
            if (line.isEmpty()) {
                continue;
            }
            widgetShown = false;
            System.out.printf("%s  · агент → LLM: запрос%s%n", GRAY, RESET);
            String answer;
            try {
                answer = agent.runTurn(sessionId, line, new OrchestratorAgent.TurnListener() {
                    @Override
                    public void onToolCall(String name, String argsJson) {
                        System.out.printf("%s  · LLM → агент: вызвать %s(%s)%s%n", GRAY, name,
                                argsJson == null ? "" : argsJson.trim(), RESET);
                    }

                    @Override
                    public void onToolResult(String name) {
                        System.out.printf("%s  · инструмент → LLM: результат %s, снова спрашиваю LLM%s%n",
                                GRAY, name, RESET);
                    }
                });
            } catch (RuntimeException e) {
                System.out.printf("%s[ошибка] %s%s%n", GRAY, e.getMessage(), RESET);
                continue;
            }
            // When a widget was shown, keep only a short comment; drop a full
            // restatement so the text doesn't duplicate the widget.
            if (!answer.isEmpty() && (!widgetShown || isBriefComment(answer))) {
                System.out.printf("%sассистент>%s %s%n", CYAN, RESET, answer);
            }
        }
    }

    /** Renders a structured widget the worker returned in an A2A DataPart. */
    static void renderWidget(Map<String, Object> w) {
        String title = str(w.get("title"));
        System.out.printf("%n%s┌─ %s%s%n", CYAN, title, RESET);
        switch (String.valueOf(w.get("_kind"))) {
            case "widget/order" -> {
                if (w.get("order") instanceof Map<?, ?> o) {
                    field("Товар:", o.get("item"));
                    field("Статус:", o.get("status_label"));
                    if (o.get("amount") != null) {
                        System.out.printf("%s│%s %-9s %s %s%n", CYAN, RESET, "Сумма:", o.get("amount"), o.get("currency"));
                    }
                    field("Клиент:", o.get("customer"));
                    field("Дата:", o.get("created"));
                    field("Ссылка:", o.get("url"));
                }
            }
            case "widget/refund_receipt" -> {
                field("Квитанция №:", w.get("receipt_id"));
                System.out.printf("%s│%s %-12s #%s — %s%n", CYAN, RESET, "Заказ:", w.get("order_id"), w.get("item"));
                if (w.get("amount") != null) {
                    System.out.printf("%s│%s %-12s %s %s%n", CYAN, RESET, "Сумма:", w.get("amount"), w.get("currency"));
                }
                if (w.get("card_last4") instanceof String last4 && !last4.isEmpty()) {
                    System.out.printf("%s│%s %-12s •••• %s%n", CYAN, RESET, "Карта:", last4);
                }
                field("Дата:", w.get("created"));
            }
            case "widget/order_list" -> {
                if (w.get("orders") instanceof List<?> rows) {
                    for (Object r : rows) {
                        if (r instanceof Map<?, ?> o) {
                            String link = o.get("url") instanceof String u && !u.isEmpty() ? "  → " + u : "";
                            System.out.printf("%s│%s #%s  %s — %s (%s %s, %s)%s%n", CYAN, RESET,
                                    o.get("id"), o.get("item"), o.get("status_label"),
                                    o.get("amount"), o.get("currency"), o.get("created"), link);
                        }
                    }
                }
            }
            default -> { // widget/confirmation and any future dialog-style widget
                String message = str(w.get("message"));
                if (!message.isEmpty()) {
                    System.out.printf("%s│%s %s%n", CYAN, RESET, message);
                }
                String labels = actionLabels(w.get("actions"));
                if (!labels.isEmpty()) {
                    System.out.printf("%s│%s %s%n", CYAN, RESET, labels);
                }
            }
        }
        System.out.printf("%s└─%s%n", CYAN, RESET);
    }

    private static void field(String label, Object v) {
        if (v != null && !String.valueOf(v).isEmpty()) {
            System.out.printf("%s│%s %-9s %s%n", CYAN, RESET, label, v);
        }
    }

    /** Renders a widget's action buttons as "[Label]  [Label]". */
    static String actionLabels(Object v) {
        if (!(v instanceof List<?> actions)) {
            return "";
        }
        StringBuilder b = new StringBuilder();
        for (Object a : actions) {
            if (a instanceof Map<?, ?> m && m.get("label") instanceof String l) {
                if (!b.isEmpty()) {
                    b.append("  ");
                }
                b.append("[").append(l).append("]");
            }
        }
        return b.toString();
    }

    /**
     * Whether an assistant reply is a short lead-in worth keeping next to a
     * widget, versus a restatement of the widget's own data.
     */
    static boolean isBriefComment(String s) {
        s = s.trim();
        if (s.isEmpty()) {
            return false;
        }
        if (s.codePointCount(0, s.length()) > 200) {
            return false;
        }
        if (s.chars().filter(c -> c == '|').count() >= 2) { // markdown table
            return false;
        }
        return s.chars().filter(c -> c == '\n').count() <= 2;
    }

    private static String str(Object o) {
        return o == null ? "" : String.valueOf(o);
    }
}
